File: CSharpCiphers/Griffinere.cs
using System.Numerics;
using System.Text;

namespace Ciphers;

public class Griffinere
{
	private readonly Dictionary<char, int> _alphabetPositionMap;
	private readonly char[] _alphabet;
	private readonly char[] _key;
	private readonly int _alphabetLength;

	public Griffinere(string key)
	{
		char[] alphabetUnique = ValidateAlphabet(GetDefaultAlphabet(), key);
		_alphabet = alphabetUnique;
		_alphabetPositionMap = CreateAlphabetPositionMap(_alphabet);
		_key = ValidateKey(key);
		_alphabetLength = _alphabet.Length;
	}

	public Griffinere(string alphabet, string key)
	{
		ArgumentException.ThrowIfNullOrWhiteSpace(alphabet);
		char[] alphabetUnique = ValidateAlphabet(alphabet, key);

		_alphabet = alphabetUnique;
		_alphabetPositionMap = CreateAlphabetPositionMap(_alphabet);
		_key = ValidateKey(key);
		_alphabetLength = _alphabet.Length;
	}

	private static char[] GetDefaultAlphabet() =>
		[
			'A','B','C','D','E','F','G','H','I','J','K','L','M','N','O','P','Q','R','S','T','U','V','W','X','Y','Z',
			'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z',
			'0','1','2','3','4','5','6','7','8','9'
		];

	/// <summary>
	/// Validates a string alphabet (unique chars, no '.'), and ensures all key chars exist in the alphabet.
	/// </summary>
	private static char[] ValidateAlphabet(string alphabet, string key)
	{
		HashSet<char> seen = new();
		List<char> uniqueList = new();

		if (alphabet.Contains('.'))
			throw new ArgumentException("Alphabet must not contain '.'");

		foreach (char c in alphabet)
		{
			if (!seen.Add(c))
				throw new ArgumentException($"Duplicate character '{c}' in provided alphabet.");
			uniqueList.Add(c);
		}

		if (uniqueList.Count < 3)
			throw new ArgumentException("Alphabet must contain at least 3 unique characters.");

		foreach (char c in key)
		{
			if (!uniqueList.Contains(c))
				throw new ArgumentException($"Alphabet does not contain the character '{c}' supplied in the key.");
		}

		return uniqueList.ToArray();
	}

	private static char[] ValidateKey(string key)
	{
		if (key.Length < 3)
			throw new ArgumentException("Key must be at least 3 characters long");

		return key.ToCharArray();
	}

	/// <summary>
	/// Validates a char[] alphabet (unique chars, no '.'), and ensures all key chars exist in the alphabet.
	/// </summary>
	private static char[] ValidateAlphabet(char[] alphabetChar, string key)
	{
		string alphabet = string.Join("", alphabetChar);
		return ValidateAlphabet(alphabet, key);
	}

	private static Dictionary<char, int> CreateAlphabetPositionMap(char[] alphabet)
	{
		Dictionary<char, int> map = new();
		for (int i = 0; i < alphabet.Length; i++)
		{
			map[alphabet[i]] = i;
		}
		return map;
	}

	private char[] GetKey(char[] encodedText)
	{
		ArgumentOutOfRangeException.ThrowIfZero(encodedText.Length);

		List<char> key = _key.ToList();
		while (key.Count < encodedText.Length)
			key.AddRange(_key);

		return key.Take(encodedText.Length).ToArray();
	}

	private char ShiftCharacterPositive(char keyChar, char textChar)
	{
		if (!TryGetCharPlacement(keyChar, out int keyPos) || !TryGetCharPlacement(textChar, out int textPos))
			return textChar; // should not trigger now that pre-encoding guarantees alphabet-only

		int shiftedIndex = (keyPos + textPos) % _alphabetLength;
		return _alphabet[shiftedIndex];
	}

	private char ShiftCharacterNegative(char keyChar, char textChar)
	{
		if (!TryGetCharPlacement(keyChar, out int keyPos) || !TryGetCharPlacement(textChar, out int textPos))
			return textChar; // should not trigger now that pre-encoding guarantees alphabet-only

		int shiftedIndex = (textPos - keyPos + _alphabetLength) % _alphabetLength;
		return _alphabet[shiftedIndex];
	}

	private bool TryGetCharPlacement(char c, out int placement) =>
		_alphabetPositionMap.TryGetValue(c, out placement);


	/// <summary>
	/// Encodes arbitrary bytes into a string using ONLY characters from the configured alphabet.
	/// Works for any alphabet length >= 2 (binary and up). Preserves leading zero bytes by
	/// prefixing the first alphabet character for each leading zero.
	/// </summary>
	private string EncodeBytesToAlphabet(ReadOnlySpan<byte> bytes)
	{
		if (bytes.Length == 0) return string.Empty;

		// Count leading zero bytes to restore later.
		int zeroCount = 0;
		for (int i = 0; i < bytes.Length && bytes[i] == 0; i++) zeroCount++;

		// BigInteger expects little-endian two's complement. Ensure positive by appending 0x00.
		byte[] littleEndian = bytes.ToArray().Reverse().Concat(new byte[] { 0 }).ToArray();
		BigInteger value = new(littleEndian);

		// Convert to base-N digits
		if (value.IsZero)
		{
			// All zeros: return zeroCount copies of alphabet[0]
			return new string(_alphabet[0], Math.Max(1, zeroCount));
		}

		List<char> encoded = new();
		BigInteger n = _alphabetLength;
		while (value > 0)
		{
			value = BigInteger.DivRem(value, n, out BigInteger rem);
			encoded.Add(_alphabet[(int)rem]);
		}

		// Append leading zeros as alphabet[0]
		for (int i = 0; i < zeroCount; i++) encoded.Add(_alphabet[0]);

		encoded.Reverse();
		return new string(encoded.ToArray());
	}

	/// <summary>
	/// Decodes a string composed ONLY of alphabet characters back into the original bytes.
	/// Restores leading zeros encoded as prefix alphabet[0] characters.
	/// </summary>
	private byte[] DecodeAlphabetToBytes(ReadOnlySpan<char> text)
	{
		if (text.Length == 0) return Array.Empty<byte>();

		// Count prefix zeros
		int zeroPrefix = 0;
		for (int i = 0; i < text.Length && text[i] == _alphabet[0]; i++) zeroPrefix++;

		BigInteger n = _alphabetLength;
		BigInteger value = BigInteger.Zero;
		foreach (char c in text)
		{
			if (!TryGetCharPlacement(c, out int digit))
				throw new ArgumentException($"Cipher text contains character '{c}' not in the alphabet.");
			value = value * n + digit;
		}

		// Convert BigInteger to big-endian bytes
		byte[] little = value.ToByteArray(); // little-endian two's complement
		if (little.Length > 1 && little[^1] == 0) // remove sign byte if present
			little = little.Take(little.Length - 1).ToArray();
		Array.Reverse(little);

		// Trim potential leading zero introduced by sign handling (not the encoded zeros)
		int trim = 0;
		while (trim < little.Length && little[trim] == 0) trim++;
		byte[] nonPadded = little.Skip(trim).ToArray();

		// Restore encoded leading zeros
		byte[] restored = new byte[zeroPrefix + nonPadded.Length];
		// zeroPrefix bytes already zero-initialized
		Buffer.BlockCopy(nonPadded, 0, restored, zeroPrefix, nonPadded.Length);
		return restored;
	}

	private char[] ToAlphabetCharArray(string text)
	{
		ArgumentException.ThrowIfNullOrWhiteSpace(text);
		byte[] bytes = Encoding.UTF8.GetBytes(text);
		string encoded = EncodeBytesToAlphabet(bytes);
		return encoded.ToCharArray();
	}

	private static string FromUtf8Bytes(byte[] bytes) => Encoding.UTF8.GetString(bytes);

	private string FromAlphabetCharArray(char[] charArray)
	{
		ArgumentOutOfRangeException.ThrowIfZero(charArray.Length);
		byte[] bytes = DecodeAlphabetToBytes(charArray);
		return FromUtf8Bytes(bytes);
	}


	public string EncryptString(string plainText)
	{
		if (string.IsNullOrWhiteSpace(plainText))
			return string.Empty;

		List<string> result = new();

		foreach (string segment in plainText.Split(' '))
		{
			if (segment == "")
			{
				result.Add(segment);
				continue;
			}

			char[] segmentChars = ToAlphabetCharArray(segment);
			char[] key = GetKey(segmentChars);
			char[] encrypted = segmentChars.Select((c, i) => ShiftCharacterPositive(key[i], c)).ToArray();
			result.Add(new string(encrypted));
		}

		string response = string.Join(" ", result);
		return response;
	}

	public string EncryptString(string plainText, int minimumResponseLength)
	{
		if (string.IsNullOrWhiteSpace(plainText))
			return string.Empty;

		if (minimumResponseLength < 1)
			throw new ArgumentOutOfRangeException(nameof(minimumResponseLength), "Minimum response length must be greater than zero.");

		string stringToFront = string.Empty;
		string stringToBack = string.Empty;

		if (plainText.Length < minimumResponseLength)
		{
			char[] charArray = plainText.Replace(" ", "").ToCharArray();
			int charArrayLength = charArray.Length;
			int needToAdd = minimumResponseLength - plainText.Length;
			int pullFromFront = (int)Math.Ceiling(Convert.ToDouble(needToAdd) / 1.25);
			int pullFromBack = needToAdd - pullFromFront;

			int addedToFront = 0;
			for (int i = 0; i < pullFromFront && addedToFront < pullFromFront; i++)
			{
				if (i < charArrayLength)
				{
					stringToFront += charArray[i];
					addedToFront++;
				}
				else
				{
					i = -1;
				}
			}

			int addedToBack = 0;
			for (int i = 0; i < pullFromBack && addedToBack < pullFromBack; i++)
			{
				if (i < charArrayLength)
				{
					stringToBack += charArray[charArray.Length - 1 - i];
					addedToBack++;
				}
				else
				{
					i = -1;
				}
			}
		}

		if (!string.IsNullOrWhiteSpace(stringToFront))
		{
			string reversedString = string.Create(stringToFront.Length, stringToFront, (span, str) =>
			{
				str.AsSpan().CopyTo(span);
				span.Reverse();
			});

			stringToFront = EncryptString(reversedString) + ".";
		}

		if (!string.IsNullOrEmpty(stringToBack))
		{
			stringToBack = "." + EncryptString(stringToBack);
		}

		string encryptedString = EncryptString(plainText);
		string response = string.Concat(stringToFront, encryptedString, stringToBack);
		return response;
	}

	public string DecryptString(string cipherText)
	{
		if (string.IsNullOrWhiteSpace(cipherText))
			return string.Empty;

		List<string> result = new();

		if (cipherText.Contains('.'))
			cipherText = cipherText.Split('.')[1];

		foreach (string segment in cipherText.Split(' '))
		{
			if (segment == string.Empty)
			{
				result.Add(string.Empty);
				continue;
			}

			char[] segmentChars = segment.ToCharArray();
			char[] key = GetKey(segmentChars);
			char[] decrypted = segmentChars.Select((c, i) => ShiftCharacterNegative(key[i], c)).ToArray();

			string decoded = FromAlphabetCharArray(decrypted);
			result.Add(decoded);
		}

		string response = string.Join(" ", result);
		return response;
	}
}

File: TypescriptCiphers/src/griffinere.ts
export class Griffinere {
	private readonly alphabet: string[];
	private readonly alphabetPositionMap: Map<string, number>;
	private readonly key: string[];
	private readonly alphabetLength: number;

	constructor(key: string, alphabet?: string) {
		const defaultAlphabet: string = Griffinere.getDefaultAlphabet();
		const validatedAlphabet: string[] = Griffinere.validateAlphabet(
			alphabet ?? defaultAlphabet,
			key,
		);

		this.alphabet = validatedAlphabet;
		this.alphabetPositionMap =
			Griffinere.createAlphabetPositionMap(validatedAlphabet);
		this.key = Griffinere.validateKey(key);
		this.alphabetLength = validatedAlphabet.length;
	}

	private static getDefaultAlphabet(): string {
		return "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
	}

	private static validateKey = (key: string): string[] => {
		if (key.length < 3) {
			throw new Error("Key must be at least 3 characters long.");
		} else {
			return key.split("");
		}
	};

	private static validateAlphabet = (
		alphabet: string,
		key: string,
	): string[] => {
		if (alphabet.includes(".")) {
			throw new Error("Alphabet must not contain '.'");
		}

		const seen: Set<string> = new Set();
		const uniqueList: string[] = [];

		for (const c of alphabet) {
			if (seen.has(c)) {
				throw new Error(
					`Duplicate character '${c}' in provided alphabet.`,
				);
			}
			seen.add(c);
			uniqueList.push(c);
		}

		if (uniqueList.length < 3) {
			throw new Error(
				"Alphabet must contain at least 3 unique characters.",
			);
		}

		for (const c of key) {
			if (!uniqueList.includes(c)) {
				throw new Error(
					`Alphabet does not contain the character '${c}' supplied in the key.`,
				);
			}
		}

		return uniqueList;
	};

	private static createAlphabetPositionMap(
		alphabet: string[],
	): Map<string, number> {
		const map: Map<string, number> = new Map();
		alphabet.forEach((char: string, index: number): void => {
			map.set(char, index);
		});
		return map;
	}

	private getKey(text: string[]): string[] {
		if (text.length === 0) throw new Error("Text cannot be empty");

		const repeatedKey: string[] = [];
		while (repeatedKey.length < text.length) {
			repeatedKey.push(...this.key);
		}
		return repeatedKey.slice(0, text.length);
	}

	private shiftCharacterPositive(keyChar: string, textChar: string): string {
		const keyPos: number | undefined =
			this.alphabetPositionMap.get(keyChar);
		const textPos: number | undefined =
			this.alphabetPositionMap.get(textChar);

		if (keyPos === undefined || textPos === undefined) return textChar; // shouldn't happen with alphabet-only input

		const shiftedIndex: number = (keyPos + textPos) % this.alphabetLength;
		return this.alphabet[shiftedIndex];
	}

	private shiftCharacterNegative(keyChar: string, textChar: string): string {
		const keyPos: number | undefined =
			this.alphabetPositionMap.get(keyChar);
		const textPos: number | undefined =
			this.alphabetPositionMap.get(textChar);

		if (keyPos === undefined || textPos === undefined) return textChar; // shouldn't happen with alphabet-only input

		const shiftedIndex: number =
			(textPos - keyPos + this.alphabetLength) % this.alphabetLength;
		return this.alphabet[shiftedIndex];
	}

	// -------------------------
	// Alphabet-only Base-N codec
	// -------------------------

	private static utf8ToBytes(text: string): Uint8Array {
		return new TextEncoder().encode(text);
	}

	private static bytesToUtf8(bytes: Uint8Array): string {
		return new TextDecoder().decode(bytes);
	}

	/**
	 * Encode arbitrary bytes into the configured alphabet (base-N).
	 * Preserves leading zero bytes by prefixing alphabet[0] for each leading zero.
	 */
	private encodeBytesToAlphabet(bytes: Uint8Array): string {
		if (bytes.length === 0) return "";

		// Count leading zero bytes
		let zeroCount = 0;
		while (zeroCount < bytes.length && bytes[zeroCount] === 0) zeroCount++;

		// Convert bytes (big-endian) to BigInt
		let value = 0n;
		for (const b of bytes) {
			value = (value << 8n) + BigInt(b);
		}

		const base = BigInt(this.alphabetLength);

		if (value === 0n) {
			// All zeros: represent at least one zero digit
			return this.alphabet[0].repeat(Math.max(1, zeroCount));
		}

		const digits: string[] = [];
		while (value > 0n) {
			const rem = Number(value % base);
			digits.push(this.alphabet[rem]);
			value = value / base;
		}

		// Add prefix zeros and reverse to big-endian digit order
		const zeroPrefix = this.alphabet[0].repeat(zeroCount);
		return zeroPrefix + digits.reverse().join("");
	}

	/**
	 * Decode text composed only of alphabet characters back to bytes.
	 * Restores leading zeros encoded as prefix alphabet[0].
	 */
	private decodeAlphabetToBytes(text: string): Uint8Array {
		if (text.length === 0) return new Uint8Array(0);

		// Count prefix zeros (alphabet[0])
		let zeroPrefix = 0;
		while (
			zeroPrefix < text.length &&
			text[zeroPrefix] === this.alphabet[0]
		) {
			zeroPrefix++;
		}

		const base = BigInt(this.alphabetLength);
		let value = 0n;
		for (const c of text) {
			const digit = this.alphabetPositionMap.get(c);
			if (digit === undefined) {
				throw new Error(
					`Cipher text contains character '${c}' not in the alphabet.`,
				);
			}
			value = value * base + BigInt(digit);
		}

		// Convert BigInt to big-endian bytes
		const bytesRev: number[] = [];
		while (value > 0n) {
			bytesRev.push(Number(value & 0xffn));
			value >>= 8n;
		}
		const core = Uint8Array.from(bytesRev.reverse());

		// Restore encoded leading zeros
		if (zeroPrefix > 0) {
			const out = new Uint8Array(zeroPrefix + core.length);
			// first zeroPrefix bytes are already zero
			out.set(core, zeroPrefix);
			return out;
		}

		return core;
	}

	private toAlphabetCharArray(text: string): string[] {
		if (!text.trim()) throw new Error("Text cannot be empty");
		const bytes: Uint8Array = Griffinere.utf8ToBytes(text);
		const encoded: string = this.encodeBytesToAlphabet(bytes);
		return encoded.split("");
	}

	private fromAlphabetCharArray(chars: string[]): string {
		if (chars.length === 0) throw new Error("Encoded input is empty");
		const text: string = chars.join("");
		const bytes: Uint8Array = this.decodeAlphabetToBytes(text);
		return Griffinere.bytesToUtf8(bytes);
	}

	// --- Public API ---

	public encryptString(plainText: string): string {
		if (!plainText.trim()) return "";

		const result: string[] = [];

		for (const segment of plainText.split(" ")) {
			if (!segment) {
				result.push(segment);
				continue;
			}

			const chars: string[] = this.toAlphabetCharArray(segment);
			const key: string[] = this.getKey(chars);
			const encrypted: string[] = chars.map(
				(c: string, i: number): string =>
					this.shiftCharacterPositive(key[i], c),
			);
			result.push(encrypted.join(""));
		}

		return result.join(" ");
	}

	public encryptStringWithMinimumLength(
		plainText: string,
		minimumLength: number,
	): string {
		if (!plainText.trim()) return "";
		if (minimumLength < 1)
			throw new Error(
				"Minimum response length must be greater than zero.",
			);

		let front = "";
		let back = "";

		if (plainText.length < minimumLength) {
			const stripped: string[] = plainText.replace(/ /g, "").split("");
			const toAdd: number = minimumLength - plainText.length;
			const frontCount: number = Math.ceil(toAdd / 1.25);
			const backCount: number = toAdd - frontCount;

			for (let i = 0; i < frontCount; i++) {
				front += stripped[i % stripped.length] ?? "";
			}
			for (let i = 0; i < backCount; i++) {
				back +=
					stripped[stripped.length - 1 - (i % stripped.length)] ?? "";
			}
		}

		const reversedFront: string = front.split("").reverse().join("");
		const frontEncrypted: string = front
			? `${this.encryptString(reversedFront)}.`
			: "";
		const backEncrypted: string = back
			? `.${this.encryptString(back)}`
			: "";
		const mainEncrypted: string = this.encryptString(plainText);

		return `${frontEncrypted}${mainEncrypted}${backEncrypted}`;
	}

	public decryptString(cipherText: string): string {
		if (!cipherText.trim()) return "";

		if (cipherText.includes(".")) {
			cipherText = cipherText.split(".")[1];
		}

		const result: string[] = [];

		for (const segment of cipherText.split(" ")) {
			if (!segment) {
				result.push("");
				continue;
			}

			const chars: string[] = segment.split("");
			const key: string[] = this.getKey(chars);
			const decrypted: string[] = chars.map(
				(c: string, i: number): string =>
					this.shiftCharacterNegative(key[i], c),
			);
			const plain: string = this.fromAlphabetCharArray(decrypted);
			result.push(plain);
		}

		return result.join(" ");
	}
}
